### ISSUE TITLE
Recursive `forwardTo` doesn't work exactly how I'd hoped

### ISSUE THREAD

**Issue by parsonsmatt** [alexmingoia/purescript-pux#14]
I'm working on a toy app with a recursive structure. Here's my [action](https://github.com/parsonsmatt/shouldistudy/blob/b6122cb41710507b603dfed98da580d114dd4796/src/Study/Pux.purs#L22-L28):

``` purescript
data Action
    = Child Int Action
    | UpdateWeight Int Number
    | UpdateScore (Score Number)
    | AddGrade
    | Undo
    | Redo
```

And an [example on how I'm using it](https://github.com/parsonsmatt/shouldistudy/blob/b6122cb41710507b603dfed98da580d114dd4796/src/Study/Pux.purs#L109-L114):

``` purescript
renderScore :: Array (Score Number) -> Html Action
renderScore gs = H.div # do
     H.button ! E.onClick (const AddGrade) # H.text "Add Grade"
     H.ul ## 
        forEachIndexed gs \i g ->
            H.li # (H.forwardTo (Child i) (viewGrade g))
```

When the UI gets nested, `forwardTo` uses the most recent function given. It'd be nice if it composed the functions instead.

**Comment by alexmingoia** [comment]
> `forwardTo` uses the most recent function given

Sorry for being thick, but I'm not sure what you mean by this. Could you explain another way? The code you posted works, yes?

**Comment by parsonsmatt** [comment]
If I have a structure like:

``` purescript
data Tree a = Tree a (Array (Tree a))

data Action = Remove | Child Int Action
```

then I can render that with the following:

``` purescript
view :: Tree String -> Html Action
view (Tree string subtrees) = ul # do
    li # do
        text string
        button ! onClick (const Remove) # text "X"
    li # ul ## forEachIndexed \i tree ->
        forwardTo (Child i) (view tree)  
```

So, if I've got a tree like:

``` purescript
example :: Tree String
example = Tree "Hello" [Tree "World" [], Tree "Party" [Tree "Adventure" [], Tree "Time" []]]
```

That should render to a list like:
- Hello [X]
  - World [X]
  - Party [X]
    - Adventure [X]
    - Time [X]

When "Adventure" has the `X` button clicked, then it'd be great for that to call `Child 1 (Child 0 Remove))`. Right now it just calls `Child 0 Remove`.

**Comment by parsonsmatt** [comment]
``` javascript
exports.mapActions = function (parentAction, html) {
  var childAction = html.props.puxParentAction;
  if (childAction) console.log(childAction, html);
  return React.cloneElement(html, {
    puxParentAction: childAction ? parentAction(childAction) : parentAction
  });
};
```

I put this `console.log` in and I'm not getting it to log anything. Will debug further

**Comment by alexmingoia** [comment]
If I understand you correctly, there's a bug in my FFI implementation since the type signature for `forwardTo` should allow this behavior?

The FFI implementation can be a bit cryptic, because instead of simply mapping over the actions, a reference to the action is saved on the tree for later use by `render` in `src/Pux.js`. This is a performance optimization to avoid additional traversals of the vtree, and the purpose of the `puxParentAction` prop. It's probably also the source of this issue.

**Comment by parsonsmatt** [comment]
The code looks like it _should_ be calling parent's parent's actions. I don't know why `puxParentAction` isn't available on the `html` parameter. I'll get a minimal example together.

**Comment by parsonsmatt** [comment]
Ok, here's an example project: https://github.com/parsonsmatt/pux-repro

![image](https://cloud.githubusercontent.com/assets/7310112/14127664/ad196fa0-f5e7-11e5-9af4-aac859cdc7d8.png)

Clicking "Time" should delete "Time", but instead it deletes "World"

![image](https://cloud.githubusercontent.com/assets/7310112/14127658/957c4cb4-f5e7-11e5-8e0e-459058c8fb87.png)

**Comment by alexmingoia** [comment]
Thanks for that! It helps a lot. For some reason it looks like `html.props.puxParentAction` isn't there even when `html` is a child view that should have been run through `fowardTo`...

**Comment by alexmingoia** [comment]
Good news! This is fixed in `v2.0.1`. Thanks for bringing this to my attention and creating that example repo.

**Comment by parsonsmatt** [comment]
Awesome, thanks! :smile:
